### PR TITLE
chore(files_versions): Use new metadata API for versions

### PR DESCRIPTION
**Pull request by artonge**
Use the new version metadata API for handling the label logic.

@emoral435 I also tweaked the API a bit, I left some comments to give some reasoning behind the changes.

As of: https://github.com/nextcloud/server/pull/44049

### CODE REVIEW

File: apps/files_versions/lib/Sabre/VersionFile.php [hunk 1]
```diff
@@ -93,18 +89,24 @@ public function setName($name) {
 		throw new Forbidden();
 	}
 
-	public function getLabel(): ?string {
```
Comment (artonge) [inline apps/files_versions/lib/Sabre/VersionFile.php:96]: Can be safely removed as this is only used in `apps/files_versions/lib/Sabre/Plugin.php` which is migrated in this PR.

File: apps/files_versions/lib/Sabre/VersionFile.php [hunk 2]
```diff
@@ -93,18 +89,24 @@ public function setName($name) {
...
-		if ($this->version instanceof INameableVersion && $this->version->getSourceFile()->getSize() > 0) {
+	public function getMetadataValue(string $key): ?string {
+		if ($this->version instanceof IMetadataVersion) {
+			return $this->version->getMetadataValue($key);
+		} elseif ($key === 'label' && $this->version instanceof INameableVersion) {
```
Comment (artonge) [inline apps/files_versions/lib/Sabre/VersionFile.php:95]: Ensure compatibility with backends that are not migrated yet.

File: apps/files_versions/lib/Versions/LegacyVersionsBackend.php [hunk 3]
```diff
@@ -314,32 +286,17 @@ private function currentUserHasPermissions(IVersion $version, int $permissions):
...
-		try {
-			$versionEntity = $this->versionsMapper->findVersionForFileId($node->getId(), $node->getMTime());
-			return $versionEntity->getMetadataValue($key);
-		} catch (\InvalidArgumentException $e) {
-			// we tried to find a version or key that doesn't exist
-			return null;
-		}
+		$versionEntity->setMetadataValue($key, $value);
+		$this->versionsMapper->update($versionEntity);
 	}
```
Comment (artonge) [inline apps/files_versions/lib/Versions/LegacyVersionsBackend.php:344]: Not needed, as any caller should be able to retrieve the metadata from the `IMetadataVersion` object.

File: apps/files_versions/lib/Versions/VersionManager.php [hunk 4]
```diff
@@ -36,7 +36,7 @@
 use OCP\IUser;
 use OCP\Lock\ManuallyLockedException;
 
-class VersionManager implements IVersionManager, INameableVersionBackend, IDeletableVersionBackend, INeedSyncVersionBackend, IMetadataVersionBackend {
```
Comment (artonge) [inline apps/files_versions/lib/Versions/VersionManager.php:39]: Involving the manager for the label or the metadata does not make sense. Caller can ask the backend directly as they have the `IVersion` object.

File: apps/files_versions/lib/Versions/Version.php [hunk 5]
```diff
@@ -124,9 +80,6 @@ public function getUser(): IUser {
 	}
 
 	public function getMetadataValue(string $key): ?string {
-		if ($this->backend instanceof IMetadataVersionBackend && $this->sourceFileInfo instanceof Node) {
-			return $this->backend->getMetadataValue($this->sourceFileInfo, "author");
-		}
-		return null;
```
Comment (artonge) [inline apps/files_versions/lib/Versions/Version.php:130]: The constructor now receives the metadata, so we can simplify the logic here.

File: apps/files_versions/lib/Sabre/VersionFile.php [hunk 6]
```diff
@@ -113,6 +106,8 @@ public function setLabel($label): bool {
 	public function getMetadataValue(string $key): ?string {
 		if ($this->version instanceof IMetadataVersion) {
 			return $this->version->getMetadataValue($key);
+		} elseif ($key === 'label' && $this->version instanceof INameableVersion) {
```
Comment (artonge) [inline apps/files_versions/lib/Sabre/VersionFile.php:109]: Ensure compatibility with backends that are not migrated yet.

File: apps/files_versions/lib/Versions/IMetadataVersionBackend.php [hunk 7]
```diff
@@ -35,18 +35,10 @@ interface IMetadataVersionBackend {
...
-
-	/**
-	 * Retrieves a corresponding value from the metadata column using the key.
-	 *
-	 * @param Node $node the node that triggered the Metadata event listener, aka, the file version
-	 * @param string $key the key for the json value of the metadata column
-	 * @since 29.0.0
-	 */
-	public function getMetadataValue(Node $node, string $key): ?string;
+	public function setMetadataValue(Node $node, int $revision, string $key, string $value): void;
```
Comment (artonge) [inline apps/files_versions/lib/Versions/IMetadataVersionBackend.php:43]: I added a `$revision` param to be able to target a specific version. @emoral435, ok with you?

**Review by emoral435 — COMMENTED** [review]
I like the new changes. The capability to store whatever metadata we want and get whichever version's metadata seems really flexible :) I do not know if this PR is ready as I have not been marked down as a reviewer, though I approve the current changes. Ping me if you need me to approve!

**Comment by artonge** [comment]
> I do not know if this PR is ready as I have not been marked down as a reviewer

I was waiting for CI to be green. Now it's ok, and your approval would be appreciated. :)

**Review by emoral435 — APPROVED** [review]

**Review by skjnldsv — APPROVED** [review]

**Pull request merged**